python/nano/src/bigdl/nano/pytorch/strategies/ray/ray_distributed.py

```diff
@@ -0,0 +1,415 @@
+                    trainer.checkpoint_callback.best_model_path, \
+                    move_data_to_device(strategy.lightning_module.state_dict(), "cpu")
+            else:
+                return move_data_to_device(results, "cpu"), \
+                    None, \
+                    move_data_to_device(strategy.lightning_module.state_dict(), "cpu")
+        else:
+            return None
+
+
+class RayStrategy(DDPSpawnStrategy):
+    """A DDP Strategy which uses ray as backend."""
+
+    strategy_name = "ray"
+
+    def __init__(self,
+                 num_workers: int = 1,
+                 num_cpus_per_worker: int = 1,
+                 use_gpu: bool = False,
+                 use_ipex: bool = False,
+                 enable_bf16: bool = False,
+                 init_hook: Callable = None,
+                 **ddp_kwargs: Any):
+        """Create a RayStrategy."""
+        # Unset MKL setting as bigdl.nano would give default values when init env.
+        # Running different programs may need different configurations.
+        # refer to https://analytics-zoo-doc.readthedocs.io/en/master/zoo.ray.html#
+        # zoo.ray.raycontext.RayServiceFuncGenerator
+        # Also KMP_AFFINITY may entangle something doing with ray._raylet.CoreWorker,
+        # which is invoked by ray.init and leads to stocking.
+        os.environ.pop("KMP_AFFINITY", None)
+        os.environ.pop("OMP_NUM_THREADS", None)
+
+        if not ray.is_initialized():    # type: ignore
+            print(ray.init())   # type: ignore
+
+        if use_ipex and TORCH_VERSION_LESS_1_10 and 'accelerator' not in ddp_kwargs:
+            super().__init__(accelerator=create_IPEXAccelerator(),
+                             parallel_devices=[],
+                             cluster_environment=RayEnvironment(world_size=num_workers),
+                             **ddp_kwargs)
+        else:
+            super().__init__(parallel_devices=[],
+                             cluster_environment=RayEnvironment(world_size=num_workers),
+                             **ddp_kwargs)
+
+        self.num_workers = num_workers
+        self.num_cpus_per_worker = num_cpus_per_worker
+        self.use_gpu = use_gpu
+        self.use_ipex = use_ipex
+        self.enable_bf16 = enable_bf16
+
+        invalidInputError(not self.use_gpu or not self.use_ipex,
+                          "You can not specify gpu and ipex at the same time.")
+
+        self.workers = self._create_worker()
+        self.init_hook = init_hook
+        self._local_rank = 0
+
+        if self.init_hook:
+            ray.get([w.execute.remote(self.init_hook) for w in self.workers])   # type: ignore
+
+    def _configure_launcher(self):
+        self._launcher = _RayLauncher(self)
+
+    def _create_worker(self):
+        """Creates Ray actor."""
+        from bigdl.nano.common.cpu_schedule import schedule_workers
+
+        cpu_procs = schedule_workers(self.num_workers)
```

Comment:
you maybe want to look at `schedule_processors` which takes into account the case where the number of workers is larger than available physical cores.